src/i18n.ts

```diff
@@ -0,0 +1,248 @@
+      source?: string;
+      isDefault?: boolean;
+      bundleOptions?: FluentBundleOptions;
+    },
+  ): Promise<void> {
+    await this.fluent.addTranslation({
+      locales: locale,
+      isDefault: locale === this.config.defaultLocale,
+      bundleOptions: this.config.fluentBundleOptions,
+      ...options,
+    });
+
+    this.locales.push(locale);
+  }
+
+  /** Get a message by its key from the specified locale. */
+  t(
+    locale: LocaleId,
+    messageId: string,
+    context?: TranslationContext,
+  ): string {
+    return this.fluent.translate(locale, messageId, context);
+  }
+
+  /** Get a message by its key from the specified locale. */
+  translate(
+    locale: LocaleId,
+    key: string,
+    context?: TranslationContext,
+  ): string {
+    return this.fluent.translate(locale, key, context);
+  }
+
+  /** Returns a middleware to .use on the `Bot` instance. */
+  middleware(): Middleware<C & I18nContextFlavor> {
+    const fluent = this.fluent;
+    const { defaultLocale, localeNegotiator, useSession } = this.config;
+    return async function (
+      ctx: C & I18nContextFlavor,
+      next: NextFunction,
+    ): Promise<void> {
+      let translate: TranslateFunction;
+
+      ctx.i18n = {
+        fluent,
+        reNegotiateLocale: negotiateLocale,
+        useLocale,
+        getLocale: getNegotiatedLocale,
+        setLocale,
+      };
+      ctx.t = translateWrapper;
+      ctx.translate = translateWrapper;
+
+      await negotiateLocale();
+      await next();
+
+      // Also exports ctx object properties for accessing them directly from
+      // the translation source files.
+      function translateWrapper(
+        messageId: string,
+        context?: TranslationContext,
+      ): string {
+        return translate(messageId, {
+          first_name: ctx.from?.first_name ?? "",
+          ctx: JSON.stringify(makeContextObject(ctx)),
+          ...context,
+        });
+      }
+
+      async function getNegotiatedLocale(): Promise<string> {
+        return await localeNegotiator?.(ctx) ??
+          // deno-lint-ignore no-explicit-any
+          (await (useSession && (ctx as any).session))?.__language_code ??
+          ctx.from?.language_code ??
+          defaultLocale;
+      }
+
+      // Determining the locale to use for translations
+      async function negotiateLocale(): Promise<void> {
+        const negotiatedLocale = await getNegotiatedLocale();
+        useLocale(negotiatedLocale);
+      }
+
+      function useLocale(locale: LocaleId): void {
+        translate = fluent.withLocale(locale);
+      }
+
+      async function setLocale(locale: LocaleId): Promise<void> {
+        if (!useSession) {
+          throw new Error(
+"You are calling `ctx.i18n.setLocale()` without setting `useSession` \
+to `true` in the configuration. It doesn't make sense because you cannot set a \
+locale in session that way. You should either enable sessions or use \
+`ctx.i18n.useLocale()` instead.",
```

Comment:
@KnorpelSenf @roj1512 Is this ok?

You are calling `ctx.i18n.setLocale()` without setting `useSession` to `true` in the configuration. It doesn't make sense because you cannot set a locale in the session that way. When you call `ctx.i18n.setLocale()` bot tries to store the user locale in the session storage. But, since you don't have session enabled, it cannot store the locale information in session storage. You should either enable sessions or use `ctx.i18n.useLocale()` instead.

Reply:
Done